Version presence absence (#36)
hdr-bgnn/Morphology-analysis: README.md

Extract morphological characteristics from image of fish trait segmentation.

> The goals of the tool is to extract presence-absence table, measurments and landmarks of fish from the segmented fish image porduced by M. Maruf.

Reviewer: ```suggestion
The goals of the tool is to extract presence-absence table, measurements and landmarks of fish from the segmented fish image produced by M. Maruf.
```

It provides a framework for creating modularized tools by using classes and jupyter notebook to visualize and test functionalities. 

> The tool is automatically containerized when a new release is published. It provide a validated version for easy integration into the [BGNN_Snakemake](https://github.com/hdr-bgnn/BGNN_Snakemake).

Reviewer: ```suggestion
The tool is automatically containerized when a new release is published. It provides a validated version for easy integration into the [BGNN_Snakemake](https://github.com/hdr-bgnn/BGNN_Snakemake).
```

> This tool can me made more generalizable but it has been developped for the [Minnows Project](https://github.com/hdr-bgnn/minnowTraits).

Reviewer: ```suggestion
This tool can me made more generalizable but it has been developed for the [Minnows Project](https://github.com/hdr-bgnn/minnowTraits).
```

## 1- Segmented image .png description

> The segmented image input looks like image below, with traits color coded and identified by "blobs". The segmentation model uses [M. Maruf's segmentation code](https://github.com/hdr-bgnn/BGNN-trait-segmentation/blob/main/Segment_mini/scripts/segmentation_main.py), and is based on a Convolutional Neural Network (CNN; more specifically unet). You can find more information on the [BGNN-trait-segementation repository](https://github.com/hdr-bgnn/BGNN-trait-segmentation).  The output is 11 classes (traits) that are color coded. We are only using 9 of them, and are excluding alt_fin_ray and caudal_fin_ray.

Reviewer: ```suggestion
The segmented image input looks like image below, with traits color coded and identified by "blobs". The segmentation model uses [M. Maruf's segmentation code](https://github.com/hdr-bgnn/BGNN-trait-segmentation/blob/main/Segment_mini/scripts/segmentation_main.py), and is based on a Convolutional Neural Network (CNN; more specifically unet). You can find more information on the [BGNN-trait-segmentation repository](https://github.com/hdr-bgnn/BGNN-trait-segmentation).  The output is 11 classes (traits) that are color coded. We are only using 9 of them, and are excluding alt_fin_ray and caudal_fin_ray.
```

* 'alt_fin_ray': [254, 102, 204],
* 'trunk': [0, 124, 124]

## 2- Presence Absence of Traits

The approach that we use for checking presence and absence of traits is the following:

  1. Isolate indivual traits (e.g., isolate the dorsal_fin)
  2. Count number of blobs

>   3. Calculate percentage of large blob

Author: Calculate percentage area of the biggest blod over the total blobs of a particular trait

Reviewer 2: will make branch to fix readme

  angle  |  fish angle using landmarks  |  Fa_lm  |  angle of the tilt of the fish from horizontal (angle between the SL_lm and and the horizontal line of the image)  
  angle  |  fish angle using PCA  |  Fa_pca  |  angle of the tilt of the fish from horizontal (angle between the pca through the midline of the fish mask and the horizontal line of the image)  
  
## 3- Extract trait information : Classes

> We created classes to add more fexibility, which can be helpfull to generalize to other projects. [Trait_class](Scripts/Traits_class.py)

Reviewer: ```suggestion
We created classes to add more flexibility, which can be helpful to generalize to other projects. [Trait_class](Scripts/Traits_class.py)
```

*Classes Overview*
+ Classes Name : Segmented_image, Measure_morphology, Visualization_morphology
+ Location : Trait_class.py
+ Description : 

>   - Segmented_image class : This class creates an object "segmented_image" from [segmented_image.png](Scripts/test_images/INHS_FISH_000742_segmented.png). The object "segmented_image" importd the image.png and split into channels corresponding to the traits (trunk, dorsal_fin, etc.) in the form of dictionnary with key = trait  and value a mask. Then multiple functions will be applied to extract basic information on individual channel and on the whole fish.

Reviewer: ```suggestion
  - Segmented_image class : This class creates an object "segmented_image" from [segmented_image.png](Scripts/test_images/INHS_FISH_000742_segmented.png). The object "segmented_image" imported the image.png and split into channels corresponding to the traits (trunk, dorsal_fin, etc.) in the form of a dictionary with key = trait  and value a mask. Then multiple functions will be applied to extract basic information on individual channel and on the whole fish.
```

    - Few functions of interest:
      + function all_landmark : find the landmarks defined in landmarks table
      + function all_measure_using_lm : measure dimension of interest using landmarks
      + fucntion all_measure_using_bbox : measure dimension of interest using bounding box

## 4- the Main function

The main function uses classes functionalities to collect information and wraps everything in final main .
The outputs are a series of .json files and .png file.
    + presence_matrix.json
    + measurements.json
    + landmark.json
    + landmark_image.png

#### 1) Using the landmarks
These trait classes functions have the suffix "_lm" to denote the method of extraction. 

The lengths (in pixels) are calculated as the distance between two landmarks (described in the "Definition" column of the trait description csvs).

#### 2) Using the bbox (bounding box)

> These trait classes fucntiosn have the suffix "_bbox" to denote the method of extraction.

Reviewer: ```suggestion
These trait classes functions have the suffix "_bbox" to denote the method of extraction.
```

The lengths (in pixels) are calculated as the distance between two landmarks (described in the "Definition" column of the trait description csvs).

_bbox (bounding box)_
These trait classes functions have the suffix "_bbox"_ to denote the method of extraction.
The lengths (in pixels) are calculated as the distance of a perpindicular line between the edges (either vertical or horizontal) of the bbox.

_mask_
These trait classes have the suffix "_m"_ to denote the method of extraction.

#### Areas

Areas are calculated as the total pixels in the mask of a trait (e.g., head area is the area of the mask of the head). These are described in the "Definition" column of the <a href="https://github.com/hdr-bgnn/Morphology-analysis/blob/main/Traits_description/Minnow_Measurements_v1.csv">Minnow_Measurements_v1.csv</a>.

## 4- Usage, input and output

By default, without option, Morphology_main.py will output the presence and absence table. Morphology, landmarks tables and landmark image are optional output. 

The outputs for the functions are a series of .json files and .png file.
    + presence_matrix.json
    + measurements.json
    + landmark.json
    + landmark_image.png

Usage:

```

> Morphology_main.py [-h] [--metadata METADATA] [--morphology MORPHOLOGY] [--landmark LANDMARK] [--lm_image LM_IMAGE] input_image output_presence 

Author: @megbalk, I think that is what you want, this is the usage description on the function with 
+ The optional arguments : [-h] [--metadata METADATA] [--morphology MORPHOLOGY] [--landmark LANDMARK] [--lm_image LM_IMAGE] 
+ the positional arguments required : input_image output_presence
